### PR TITLE
JIT: Add a primitive to split statements at a specified tree

### PR DESCRIPTION
**Pull request by jakobbotsch**
Add a `gtSplitTree` that can split a statement at an arbitrary tree: this creates new statements before the statement such that the original statement now executes the input tree as the very first non-invariant node.

This can be used for example to introduce control flow at an arbitrary point within a tree.

### CODE REVIEW

**Comment by BruceForstall** [comment]
It seems totally reasonable to add this as one of the JIT stress modes if it is incorporated into the JIT.

Could we use gtSplitTree to remove all GT_COMMA by hoisting the op1 to its own statement? There are probably lots of optimizations that are doing tree-based pattern matching including commas that would generate worse code (or fail) as a result, but maybe as a stress mode?

**Comment by jakobbotsch** [comment]
> It seems totally reasonable to add this as one of the JIT stress modes if it is incorporated into the JIT.
> 
> Could we use gtSplitTree to remove all GT_COMMA by hoisting the op1 to its own statement? There are probably lots of optimizations that are doing tree-based pattern matching including commas that would generate worse code (or fail) as a result, but maybe as a stress mode?

Yes, I think I will end up PR'ing this primitive separately from #81635, and it seems like a good idea to have these stress modes. Indeed it can be used to remove all COMMA nodes, that would also be an interesting stress mode, I'll add it once this is in a more workable state.

File: src/coreclr/jit/compiler.cpp [hunk 1]
```diff
@@ -5337,6 +5340,157 @@ PhaseStatus Compiler::placeLoopAlignInstructions()
...
+
+    return PhaseStatus::MODIFIED_NOTHING;
+}
+
+//------------------------------------------------------------------------
+// SplitTreesRandomly: Split all statements at a random location.
+//
+void Compiler::SplitTreesRandomly()
+{
+    CLRRandom rng;
```
Comment (jakobbotsch) [inline src/coreclr/jit/compiler.cpp:5374]:
```suggestion
#ifdef DEBUG
    CLRRandom rng;
```

File: src/coreclr/jit/compiler.cpp [hunk 2]
```diff
@@ -5337,6 +5340,157 @@ PhaseStatus Compiler::placeLoopAlignInstructions()
...
+                        fgMorphStmtBlockOps(block, stmt);
+                    }
+
+                    break;
+                }
+
+                splitTree--;
+            }
+        }
+    }
```
Comment (jakobbotsch) [inline src/coreclr/jit/compiler.cpp:5420]:
```suggestion
    }
#endif
```

**Comment by jakobbotsch** [comment]
> There are probably lots of optimizations that are doing tree-based pattern matching including commas that would generate worse code (or fail) as a result, but maybe as a stress mode?

In case you are curious, here are the diffs if we remove all commas before the optimization phases:
https://gist.github.com/jakobbotsch/e210230f6d154c7547ae0823ed32b230

**Comment by BruceForstall** [comment]
> In case you are curious, here are the diffs

As expected... bad :-)  If optimizations used SSA instead of tree-based pattern matching, maybe it wouldn't be so bad.

**Comment by jakobbotsch** [comment]
Previous stress runs were clean except for #83298. I had to merge this to pick up #83224 that this was based on, but that shouldn't require a rerun of those pipelines.

cc @dotnet/jit-contrib PTAL @EgorBo

File: src/coreclr/jit/compiler.cpp [hunk 3]
```diff
@@ -5336,6 +5339,163 @@ PhaseStatus Compiler::placeLoopAlignInstructions()
...
+                    {
+                        while ((newStmt != nullptr) && (newStmt != stmt))
+                        {
+                            fgMorphStmtBlockOps(block, newStmt);
+                            newStmt = newStmt->GetNextStmt();
+                        }
+
+                        fgMorphStmtBlockOps(block, stmt);
+                        gtUpdateStmtSideEffects(stmt);
+                    }
```
Comment (jakobbotsch) [inline src/coreclr/jit/compiler.cpp:5413]: @EgorBo Your PR will need to do something similar like this, but note that calling `fgMorphStmtBlockOps` can invalidate the use, so it should be done only after you've replaced it with the new local.

File: src/coreclr/jit/fgstmt.cpp [hunk 4]
```diff
@@ -18,7 +18,7 @@ bool Compiler::fgBlockContainsStatementBounded(BasicBlock* block,
                                                Statement*  stmt,
                                                bool        answerOnBoundExceeded /*= true*/)
 {
-    const __int64 maxLinks = 1000000000;
+    const __int64 maxLinks = 100000000;
```
Comment (jakobbotsch) [inline src/coreclr/jit/fgstmt.cpp:21]: I thought this limit was a bit on the high side before.

**Review by EgorBo — APPROVED** [review]
Thanks! I've been following changes in this PR so LGTM

**Comment by jakobbotsch** [comment]
> As expected... bad :-) If optimizations used SSA instead of tree-based pattern matching, maybe it wouldn't be so bad.

I realized while working on #89560 that the comma removal is more conservative than necessary -- it introduces a local for the comma's op2, which is unnecessary. 

With that fixed the diffs look like https://gist.github.com/jakobbotsch/dd2ea6d7b94d9dbf80283191d98e2dd0, so it goes from "very bad" to "still very bad, but only half as very bad".

**Comment by EgorBo** [comment]
Wonder what happens in this small guy: `System.Numerics.Vector2:get_Item(int):float:this (FullOpts)`
image

**Pull request merged**